### PR TITLE
Add EnableRangeProcessing

### PR DESCRIPTION
**Pull request by jbagga**
Addresses #6780 

Ensured that this change solves the multiple GET requests issue described in #6780.

@Tratcher @rynowak

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/FileResultExecutorBase.cs [hunk 1]
```diff
@@ -41,6 +41,7 @@ internal enum PreconditionState
             ActionContext context,
             FileResult result,
             long? fileLength,
+            bool enableRangeProcessing,
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/FileResultExecutorBase.cs]: Is .Infrastructure considered public API? the class is public and the method is protected, so people could derive from it.

Comment (jbagga) [reply src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/FileResultExecutorBase.cs]: It was .Internal until recently.

@Eilon is Infrastructure just a name change and is considered internal still?

**Comment by jbagga** [comment]
@Eilon since .Infrastructure is a public namespace, this is a breaking change

File: src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/FileResultExecutorBase.cs (hunk 1 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/FileResultExecutorBase.cs]: @jbagga was this in `.Infrastructure` in 2.0.0? That's the salient point

If this class was in `.Internal` then it has no support bar in 2.0.0 and there is no breaking change 😆

Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/FileResultExecutorBase.cs]: Or rather, it's a change that we're willing to make

Comment (jbagga) [reply src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/FileResultExecutorBase.cs]: It was .Internal in 2.0.0 https://github.com/aspnet/Mvc/blob/rel/2.0.0/src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs#L16

**Comment by spallister** [comment]
@jbagga - Could you please confirm if range processing, to enable 206 Partial Content responses, will not be in ASP.NET Core until 2.1, which has yet to be released?  Reason I'm asking is because the online documentation [here](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.controllerbase.file?view=aspnetcore-2.0#Microsoft_AspNetCore_Mvc_ControllerBase_File_System_String_System_String_) for ASP.NET Core 2.0 - which I'm currently using - seems to imply that this is already implemented:

> Returns the file specified by virtualPath (Status200OK) with the specified contentType as the Content-Type. **This supports range requests (Status206PartialContent or Status416RangeNotSatisfiable if the range is not satisfiable).**

Intellisense within Visual Studio is similarly mentioning support for range requests when working with File Result Executors (e.g. `ControllerBase.File`), however, the action always returns 200 OK results, and correspondingly, the `Content-Range` header is never returned in the response.

Thanks in advance for your help!

**Comment by jbagga** [comment]
@spallister It should work for 2.0.0 by default. Can you share a repro and the exact SDK you are using? In 2.0.x patch we added a switch to turn on range processing https://github.com/aspnet/Mvc/issues/6792 (not ON by default).

**Comment by spallister** [comment]
@jbagga - I was using ASP.NET Core 2.0.3, then just tried updating to ASP.NET Core 2.0.5, and see that I'm using 2.1.4 of the .NET Core SDK upon running `dotnet --version` in the Package Manager Console of Visual Studio - same result.

Here's the action that I'm trying to execute:
```c#
/// <summary>
/// Get Video File by ID.
/// </summary>
[HttpGet("{id}/file")]
public async Task<FileStreamResult> GetVideoFileById(Guid id)
{
    var query = new GetVideoFilePathQuery(id);

    var videoFilePathDto = await _mediator.Send(query);

    if (videoFilePathDto != null)
    {
        return File(System.IO.File.OpenRead(videoFilePathDto.FilePath), videoFilePathDto.FileContentType);
    }
    else
    {
        return null;
    }
}
```

I was able to successfully return 206 Partial Content responses using a SO contributor's VideoStreamResult implementation [here](https://stackoverflow.com/a/35920244/2391336) (e.g. replaced `FileStreamResult` and `File` with `VideoStreamResult` and `new VideoStreamResult` correspondingly, in the above).

For the switch you mention in your above post, how do I go about enabling it (i.e. where is it located)?

**Comment by jbagga** [comment]
You can set the switch like [this](https://github.com/Eilon/QuirksModeSwitchSample/blob/master/QuirksModeSwitchSample/runtimeconfig.template.json) example in a runtimeconfig.json file for .NET Core

```
{
  // Set the switch here to affect .NET Core apps
  "configProperties": {
    "Switch.Microsoft.AspNetCore.Mvc.EnableRangeProcessing": "true"
  }
}

```

This will turn on range processing for `FileStreamResult` and `FileContentResult` for 2.0.x. If this does not solve your problem, please open an issue with a repro and I will investigate further!

**Comment by spallister** [comment]
That's got it - Thanks, @jbagga!

I ended up setting the switch programmatically in Program.cs:

```c#
// Enable 206 Partial Content responses to enable Video Seeking from api/videos/{id}/file,
// as per, https://github.com/aspnet/Mvc/pull/6895#issuecomment-356477675.
// Should be able to remove this switch and use the enableRangeProcessing overload of File once 
// ASP.NET Core 2.1 released
AppContext.SetSwitch("Switch.Microsoft.AspNetCore.Mvc.EnableRangeProcessing", true);
```

**Pull request merged**